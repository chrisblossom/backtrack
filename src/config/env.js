/* @flow */

const env = {
    productionRelease: process.env.PRODUCTION_RELEASE !== 'false',

    development: process.env.NODE_ENV
        ? process.env.NODE_ENV === 'development'
        : true,

    /**
     * HMR enabled by default
     *
     * Always disable HMR in production-like environments
     */
    hmr:
        process.env.RUN_MODE === 'dev' && process.env.HMR !== 'false'
            ? process.env.NODE_ENV !== 'production'
            : false,
};

module.exports = Object.freeze(env);
