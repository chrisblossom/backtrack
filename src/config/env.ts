const productionRelease = process.env.PRODUCTION_RELEASE !== 'false';

const development = process.env.NODE_ENV
	? process.env.NODE_ENV === 'development'
	: true;

/**
 * HMR enabled by default
 *
 * Always disable HMR in production-like environments
 */
const hmr =
	process.env.RUN_MODE === 'dev' && process.env.HMR !== 'false'
		? process.env.NODE_ENV !== 'production'
		: false;

export { productionRelease, development, hmr };
