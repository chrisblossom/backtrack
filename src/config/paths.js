/* @flow */

import { realpathSync } from 'fs';
import path from 'path';

/**
 * Make sure any symlinks in the project folder are resolved:
 * https://github.com/facebookincubator/create-react-app/issues/637
 */
const rootPath = realpathSync(process.cwd());

const paths = {
    rootPath,
    buildPath: path.resolve(rootPath, 'dist'),
    sourcePath: path.resolve(rootPath, 'src'),
};

module.exports = Object.freeze(paths);
