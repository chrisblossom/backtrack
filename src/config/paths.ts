import { realpathSync } from 'fs';
import path from 'path';

/**
 * Make sure any symlinks in the project folder are resolved:
 * https://github.com/facebookincubator/create-react-app/issues/637
 */
const rootPath = realpathSync(process.cwd());
const buildPath = path.resolve(rootPath, 'dist');
const sourcePath = path.resolve(rootPath, 'src');

export { rootPath, buildPath, sourcePath };
