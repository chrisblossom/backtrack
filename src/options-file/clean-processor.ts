import { NormalizedClean } from '../types';

/**
 * Patterns are only allowed inside the buildPath.
 */
const base = {
	del: [],
	makeDirs: [],
	copy: [],
};

interface Args {
	value: NormalizedClean[];
	current?: NormalizedClean;
}

function cleanProcessor({ value, current = base }: Args) {
	return value.reduce(
		(acc, arg) => {
			arg.del.forEach((pattern) => {
				/**
				 * remove duplicates
				 */
				if (acc.del.includes(pattern) === false) {
					acc.del.push(pattern);
				}
			});

			arg.makeDirs.forEach((dir) => {
				/**
				 * remove duplicates
				 */
				if (acc.makeDirs.includes(dir) === false) {
					acc.makeDirs.push(dir);
				}
			});

			arg.copy.forEach((copy) => {
				/**
				 * remove duplicates
				 */
				const isDuplicate = !!acc.copy.find((copyArg) => {
					return (
						copyArg.src === copy.src && copyArg.dest === copy.dest
					);
				});

				if (isDuplicate === false) {
					acc.copy.push(copy);
				}
			});

			return acc;
		},
		{
			del: [...current.del],
			makeDirs: [...current.makeDirs],
			copy: [...current.copy],
		},
	);
}

export { cleanProcessor };
