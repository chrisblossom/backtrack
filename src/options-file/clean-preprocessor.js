/* @flow */

import path from 'path';
import { toArray } from '../utils/object-utils';
import { buildPath, rootPath } from '../config/paths';
import type { Clean, NormalizedClean } from '../types.js';

// eslint-disable-next-line flowtype/require-exact-type
type Args = {
    value?: Clean | $ReadOnlyArray<Clean>,
};

function cleanPreprocessor({ value }: Args = {}): $ReadOnlyArray<
    NormalizedClean,
> {
    const allToArray = toArray(value);

    return allToArray.map((arg: Clean) => {
        const makeDirs = toArray(arg.makeDirs).map((makeDir) => {
            const resolvedMakeDir = path.resolve(buildPath, makeDir);
            return resolvedMakeDir;
        });

        const copy = toArray(arg.copy).map((copyArgs) => {
            const resolvedSrc = path.resolve(rootPath, copyArgs.src);
            const resolvedDest = path.resolve(buildPath, copyArgs.dest);

            const result = {
                src: resolvedSrc,
                dest: resolvedDest,
                hash: copyArgs.hash,
            };

            if (result.hash === undefined) {
                delete result.hash;
            }

            return result;
        });

        return {
            del: toArray(arg.del),
            makeDirs,
            copy,
        };
    });
}

export { cleanPreprocessor };
