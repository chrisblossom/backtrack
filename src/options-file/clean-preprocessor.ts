import path from 'path';
import { toArray } from '../utils/object-utils';
import { buildPath, rootPath } from '../config/paths';
import { Clean, NormalizedClean } from '../types';

interface Args {
	value?: Clean | Clean[];
}

function cleanPreprocessor({ value }: Args = {}): NormalizedClean[] {
	const allToArray = toArray(value);

	return allToArray.map((arg: Clean) => {
		const makeDirs = toArray(arg.makeDirs).map((makeDir) => {
			const resolvedMakeDir = path.resolve(buildPath, makeDir);
			return resolvedMakeDir;
		});

		const copy = toArray(arg.copy).map((copyArgs) => {
			const resolvedSrc = path.resolve(rootPath, copyArgs.src);
			const resolvedDest = path.resolve(buildPath, copyArgs.dest);

			const result = {
				src: resolvedSrc,
				dest: resolvedDest,
				hash: copyArgs.hash,
			};

			if (result.hash === undefined) {
				delete result.hash;
			}

			return result;
		});

		return {
			del: toArray(arg.del),
			makeDirs,
			copy,
		};
	});
}

export { cleanPreprocessor };
