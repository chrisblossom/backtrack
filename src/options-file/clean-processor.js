/* @flow */

import path from 'path';
import { buildPath } from '../config/paths';

import type { NormalizedClean } from '../types.js';

/**
 * Patterns are only allowed inside the buildPath.
 */
const base = {
    del: [],
    makeDirs: [],
};

// eslint-disable-next-line flowtype/require-exact-type
type Args = {
    value: $ReadOnlyArray<NormalizedClean>,
    current?: NormalizedClean,
};

function cleanProcessor({ value, current = base }: Args) {
    return value.reduce(
        (acc, arg) => {
            arg.del.forEach((pattern) => {
                /**
                 * remove duplicates
                 */
                if (acc.del.includes(pattern) === false) {
                    acc.del.push(pattern);
                }
            });

            arg.makeDirs.forEach((dir) => {
                const resolvedDir = path.resolve(buildPath, dir);

                /**
                 * remove duplicates
                 */
                if (acc.makeDirs.includes(resolvedDir) === false) {
                    acc.makeDirs.push(resolvedDir);
                }
            });

            return acc;
        },
        {
            del: [...current.del],
            makeDirs: [...current.makeDirs],
        },
    );
}

export { cleanProcessor };
