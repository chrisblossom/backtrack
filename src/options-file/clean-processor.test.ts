import path from 'path';
import { buildPath, rootPath } from '../config/paths';

require('./clean-processor');
require('./clean-preprocessor');

const cleanProcessor = (args: any) =>
    require('./clean-processor').cleanProcessor(args);

const cleanPreprocessor = (args: any) =>
    require('./clean-preprocessor').cleanPreprocessor(args);

function getValue(value: any) {
    return cleanPreprocessor({ value });
}

describe('cleanProcessor', () => {
    test('merges args together and completes resolves path', () => {
        const current = {
            del: ['first', 'second'],
            makeDirs: [
                path.resolve(buildPath, 'first'),
                path.resolve(buildPath, 'second'),
            ],

            copy: [
                {
                    src: path.resolve(rootPath, 'static-1'),
                    dest: path.resolve(buildPath, 'static-1'),
                },
            ],
        };

        const value = getValue([
            {
                del: ['third'],
                makeDirs: ['third'],
                copy: [
                    {
                        src: 'static-2',
                        dest: 'static-2',
                    },
                    {
                        src: 'static-3',
                        dest: 'static-3',
                        hash: true,
                    },
                ],
            },
        ]);

        const result = cleanProcessor({ current, value });

        expect(result).toMatchSnapshot();
    });

    test('removes duplicates', () => {
        const value = getValue([
            {
                del: ['first'],
                makeDirs: ['first'],
                copy: [
                    {
                        src: 'static-1',
                        dest: 'static-1',
                    },
                ],
            },
            {
                del: ['first'],
                makeDirs: ['first'],
                copy: [
                    {
                        src: 'static-1',
                        dest: 'static-1',
                    },
                    {
                        src: 'static-1',
                        dest: 'static-1',
                        hash: true,
                    },
                ],
            },
        ]);

        const result = cleanProcessor({ value });

        expect(result).toMatchSnapshot();
    });
});
