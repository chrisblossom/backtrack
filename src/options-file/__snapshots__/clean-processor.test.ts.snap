// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cleanProcessor merges args together and completes resolves path 1`] = `
Object {
  "copy": Array [
    Object {
      "dest": "<PROJECT_ROOT>/dist/static-1",
      "src": "<PROJECT_ROOT>/static-1",
    },
    Object {
      "dest": "<PROJECT_ROOT>/dist/static-2",
      "src": "<PROJECT_ROOT>/static-2",
    },
    Object {
      "dest": "<PROJECT_ROOT>/dist/static-3",
      "hash": true,
      "src": "<PROJECT_ROOT>/static-3",
    },
  ],
  "del": Array [
    "first",
    "second",
    "third",
  ],
  "makeDirs": Array [
    "<PROJECT_ROOT>/dist/first",
    "<PROJECT_ROOT>/dist/second",
    "<PROJECT_ROOT>/dist/third",
  ],
}
`;

exports[`cleanProcessor removes duplicates 1`] = `
Object {
  "copy": Array [
    Object {
      "dest": "<PROJECT_ROOT>/dist/static-1",
      "src": "<PROJECT_ROOT>/static-1",
    },
  ],
  "del": Array [
    "first",
  ],
  "makeDirs": Array [
    "<PROJECT_ROOT>/dist/first",
  ],
}
`;
